src/main/views/pluginsPopUps.ts

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import type {
+    BrowserWindow,
+    Event,
+    WebContentsWillNavigateEventParams,
+    WebContentsWillRedirectEventParams,
+    WebContentsDidStartNavigationEventParams,
+} from 'electron';
+import {shell} from 'electron';
+
+import ServerViewState from 'app/serverViewState';
+import {Logger} from 'common/log';
+import {
+    isTeamUrl,
+    parseURL,
+} from 'common/utils/url';
+import ContextMenu from 'main/contextMenu';
+import ViewManager from 'main/views/viewManager';
+import {generateHandleConsoleMessage} from 'main/views/webContentEventsCommon';
+import MainWindow from 'main/windows/mainWindow';
+
+const log = new Logger('PluginsPopUpsManager');
+
+type PluginPopUp = {
+    parentId: number;
+    win: BrowserWindow;
+}
+
+export class PluginsPopUpsManager {
+    popups: Record<number, PluginPopUp>;
+
+    constructor() {
+        this.popups = {};
+    }
+
+    generateHandleCreateWindow = (parentId: number) => (win: BrowserWindow, details: Electron.DidCreateWindowDetails) => {
+        const webContentsId = win.webContents.id;
+
+        log.debug('created popup window', details.url, webContentsId);
+        this.popups[webContentsId] = {
+            parentId,
+            win,
+        };
+
+        // We take a conservative approach for the time being and disallow most events coming from popups:
+        // - Redirects
+        // - Navigation
+        // - Opening new windows
+        win.webContents.on('will-redirect', (ev: Event<WebContentsWillRedirectEventParams>) => {
+            log.warn(`prevented popup window from redirecting to: ${ev.url}`);
+            ev.preventDefault();
+        });
+        win.webContents.on('will-navigate', (ev: Event<WebContentsWillNavigateEventParams>) => {
+            if (ev.url === details.url) {
+                return;
+            }
+
+            log.warn(`prevented popup window from navigating to: ${ev.url}`);
+            ev.preventDefault();
+        });
+        win.webContents.on('did-start-navigation', (ev: Event<WebContentsDidStartNavigationEventParams>) => {
+            if (ev.url === details.url) {
+                return;
+            }
+
+            log.warn(`prevented popup window from navigating to: ${ev.url}`);
+            ev.preventDefault();
+        });
+        win.webContents.setWindowOpenHandler(({url}): {action: 'deny' | 'allow'} => {
```

Comment:
Just so I understand, is it intended that this function never returns {action: 'allow'}?

Reply:
Yes, we prevent popups from opening more child windows. We still want to allow to open links to either browser (external) or redirect in the main app (e.g. if it's a link to a post/channel).

Reply:
@enzowritescode That's just to make typescript happy. Since we never return `allow` I suppose we could remove that.

> Also, are there any other attributes in the details object that we should be checking aside from URL?

You tell me :p We are denying and only using the URL so we don't really care about the rest I suppose. This is not differing from how we handle this event in other places.

Reply:
@streamer45 I think we should remove `allow`. 
I will have to look further into the other stuff to see if it's of any use.

Reply:
Done